File: examples/user_notification_java/src/main/java/net/codecrete/windowsapi/examples/UserNotificationCallback.java
//
// Windows API Generator for Java
// Copyright (c) 2025 Manuel Bleichenbacher
// Licensed under MIT License
// https://opensource.org/licenses/MIT
//
package net.codecrete.windowsapi.examples;

import windows.win32.foundation.POINT;
import windows.win32.system.com.IUnknown;
import windows.win32.ui.shell.IUserNotificationCallback;

import java.lang.foreign.MemorySegment;

import static windows.win32.foundation.Constants.S_OK;

/**
 * Implementation of COM interface `IUserNotificationCallback`
 */
public class UserNotificationCallback extends ComObject implements IUserNotificationCallback {
    private static final MemorySegment[] implementedIIDs
            = new MemorySegment[]{IUnknown.iid(), IUserNotificationCallback.iid()};

    public UserNotificationCallback() {
        super(implementedIIDs);
    }

    @Override
    public int OnBalloonUserClick(MemorySegment pt) {
        printAction("clicked balloon", pt);
        return S_OK;
    }

    @Override
    public int OnLeftClick(MemorySegment pt) {
        printAction("clicked icon", pt);
        return S_OK;
    }

    @Override
    public int OnContextMenu(MemorySegment pt) {
        printAction("triggered context menu", pt);
        return S_OK;
    }

    @SuppressWarnings("java:S106")
    private void printAction(String action, MemorySegment point) {
        var x = POINT.x(point);
        var y = POINT.y(point);
        System.out.printf("%s at (%d, %d)%n", action, x, y);
    }
}

File: examples/user_notification_kotlin/src/main/kotlin/net/codecrete/windowsapi/examples/UserNotificationCallback.kt
//
// Windows API Generator for Java
// Copyright (c) 2025 Manuel Bleichenbacher
// Licensed under MIT License
// https://opensource.org/licenses/MIT
//
package net.codecrete.windowsapi.examples

import windows.win32.foundation.Constants.S_OK
import windows.win32.foundation.POINT
import windows.win32.system.com.IUnknown
import windows.win32.ui.shell.IUserNotificationCallback
import java.lang.foreign.MemorySegment

private val implementedIIDs = arrayOf(IUnknown.iid(), IUserNotificationCallback.iid())

/**
 * Implementation of COM interface `IUserNotificationCallback`
 */
class UserNotificationCallback : ComObject(implementedIIDs), IUserNotificationCallback {
    override fun OnBalloonUserClick(pt: MemorySegment): Int {
        printAction("clicked balloon", pt)
        return S_OK
    }

    override fun OnLeftClick(pt: MemorySegment): Int {
        printAction("clicked icon", pt)
        return S_OK
    }

    override fun OnContextMenu(pt: MemorySegment): Int {
        printAction("triggered context menu", pt)
        return S_OK
    }

    private fun printAction(action: String, point: MemorySegment) {
        val x = POINT.x(point)
        val y = POINT.y(point)
        println("$action at ($x, $y)")
    }
}